### PR TITLE
SDK: refactor builder script

### PR DESCRIPTION
**Pull request by simison**
- Import builder as a function instead of calling it via command line

- Adds instructions for both bin and run -scripts and an example. Instructions are dynamic depending if you run the script via `npx`/`yarn` or `npm run`

- Adds `--mode=production|development` arg

- Better logging, inspired by https://webpack.js.org/api/node/#error-handling

### From Calypso
```bash
npm run sdk:gutenberg -- --editor-script=client/gutenberg/extensions/hello-dolly/hello-block.js
```
Note extra `--` delimiter because that's how npm can pass arguments down to the sdk script. Read more: https://docs.npmjs.com/cli/run-script

This also works:
```bash
./bin/sdk-cli.js gutenberg --editor-script=client/gutenberg/extensions/hello-dolly/hello-block.js
```

You should end up with `editor-script.js` file in `client/gutenberg/extensions/hello-dolly/build/`

See how help instructions use "npm run":
```
npm run sdk:gutenberg --help
```

### From Jetpack folder:

```bash
npx calypso-sdk gutenberg --editor-script=client/gutenberg/extensions/hello-dolly/hello-block.js --output-dir=./hello-dolly
```

You should end up with `editor-script.js` file in `./hello-dolly`

See how help instructions use "calypso-sdk":
```
npm run calypso-sdk --help
```

Test also adding `--mode=development` to not get your JS minified.

### CODE REVIEW

**Comment by simison** [comment]
Now now! We're back to baking flag SVGs in the build... :-( 

Would love to come up with something else than:
```js
process.env[ 'SKIP_FLAG_IMAGES' ] = 'true';`
```

File: bin/sdk-cli.js [hunk 1]
```diff
@@ -10,41 +10,51 @@ const path = require( 'path' );
...
-		shell: true,
-		stdio: 'inherit',
-	} );
-};
+// Script name is used in help instructions;
+// pick between `npm run calypso-sdk` and `npx calypso-sdk`.
+// Show also how npm scripts require delimiter to pass arguments.
+const calleeScript = path.basename( process.argv[ 1 ] );
+const scriptName = calleeScript === path.basename( __filename ) ? 'npm run calypso-sdk' : calleeScript;
+const delimit = scriptName.substring( 0, 3 ) === 'npm' ? '-- ' : '';
```
Comment (simison) [inline bin/sdk-cli.js]: Not sure if this is overcomplicated at this point but since we don't have any documentation, having clear `--help` instructions is important.

File: bin/sdk/gutenberg.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+const chalk = require( 'chalk' );
+const path = require( 'path' );
+const webpack = require( 'webpack' );
+
+const __rootDir = path.resolve( __dirname, '../../' );
+const baseConfig = require( path.join( __rootDir, 'webpack.config.js' ) );
+
+exports.compile = args => {
+
+	const options = {
```
Comment (simison) [inline bin/sdk/gutenberg.js]: It's unclear now that these are required:
- `args.mode`: CLI script defaults this to `production`. Defaulting could be moved to this file but it's nice to have CLI indicate default with `--help` 
- `editorScript`: CLI script requires this

File: bin/sdk/gutenberg.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+		},
+	};
+
+	const compiler = webpack( config );
+
+	compiler.run( ( error, stats ) => {
+		if ( error ) {
+			console.error( error );
+			console.log( chalk.red( 'Failed to build Gutenberg extension' ) );
+			process.exit( 1 );
```
Comment (simison) [inline bin/sdk/gutenberg.js]: Without this, the process was exiting cleanly with `0`

File: bin/sdk/gutenberg.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+				wp: 'wp',
+			},
+			optimization: {
+				splitChunks: false,
+			},
+			output: {
+				path: options.outputDir,
+				filename: `[name].js`,
+				libraryTarget: 'window',
+				library: `gutenberg-extension-[name]`,
```
Comment (simison) [inline bin/sdk/gutenberg.js]: It's a bit unclear if we actually need this with `libraryTarget: 'window'`?

https://webpack.js.org/guides/author-libraries/

File: bin/sdk-cli.js [hunk 5]
```diff
@@ -10,41 +10,51 @@ const path = require( 'path' );
...
+	.scriptName( scriptName )
+	.usage( `Usage: $0 <command> ${ delimit }[options]` )
+	.example( `$0 build-gutenberg ${ delimit }--editor-script=hello-dolly.js` )
 	.command( {
-		command: 'build-block',
-		desc: 'Build a block',
+		command: 'build-gutenberg',
+		desc: 'Build a Gutenberg extension',
 		builder: yargs => yargs.options( {
+			'mode': {
```
Comment (simison) [inline bin/sdk-cli.js]: Noting that `mode` arg could as well be in a separate PR to focus this one a bit more since PR has plenty going on already.

File: bin/sdk/gutenberg.js (hunk 4 above)
Comment (simison) [reply bin/sdk/gutenberg.js]: I just tested without it and hello-dolly worked just fine 🤷‍♂️ 

It would matter more if we make `libraryTarget` configurable

Comment (simison) [reply bin/sdk/gutenberg.js]: Oh and _with_ `library`, we really need to ensure they're unique. Otherwise different files can end up overwriting each other:
```js
window['gutenberg-extension-editor-script']=function
```

So I'd drop it out to keep things more simple for now.

**Comment by tyxla** [comment]
> Now now! We're back to baking flag SVGs in the build... :-(
> Would love to come up with something else than:
> `process.env[ 'SKIP_FLAG_IMAGES' ] = 'true';`

I've suggested an env-var-less alternative here: #26350

**Comment by simison** [comment]
Rebased now that we've renamed some things and we're no more blocked by flag SVGs.
